oak_services/src/http/mod.rs

```diff
@@ -0,0 +1,105 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::proto::oak::encap::{HeaderMap, HeaderValue, HttpRequest, HttpResponse};
+use http::{Request, Response};
+use log::warn;
+use maplit::hashmap;
+use oak_abi::OakStatus;
+use std::convert::TryInto;
+
+impl From<Response<Vec<u8>>> for HttpResponse {
+    fn from(response: Response<Vec<u8>>) -> Self {
+        HttpResponse {
+            body: response.body().to_owned(),
+            status: response.status().as_u16() as i32,
+            headers: Some(HeaderMap::from(response.headers().to_owned())),
+        }
+    }
+}
+
+impl TryInto<Request<Vec<u8>>> for HttpRequest {
+    type Error = OakStatus;
+
+    fn try_into(self) -> Result<Request<Vec<u8>>, Self::Error> {
+        let mut builder = Request::builder()
+            .method(self.method.as_bytes())
+            .uri(self.uri);
+        if let Some(headers) = self.headers {
+            for (header_name, header_value) in headers.to_vec() {
+                builder = builder.header(header_name, header_value);
+            }
+        }
+        builder.body(self.body).map_err(|err| {
+            warn!("Could not create body from vector: {}", err);
```

Comment:
I don't think an error here means there is a problem with the body, in fact most likely it will be in the headers, which need to follow specific rules and escaping etc. So I think the message should be about creating the request, not just the body.

Also see https://docs.rs/http/0.2.1/http/response/struct.Builder.html#method.body:

> This function may return an error if any previously configured argument failed to parse or get converted to the internal representation. For example if an invalid head was specified via header("Foo", "Bar\r\n") the error will be returned when this function is called rather than when header was called.